File: hackerrank/problem solving/algorithms/2 - implementation/26 - circular array rotation/with morfo library/circular_array_rotation.py
# https://www.hackerrank.com/challenges/circular-array-rotation/problem?isFullScreen=true

import morfo


def main():
    n, n_rotation, n_queries = morfo.read(int, 3)
    arr = morfo.read(int, n)
    print(*queries(arr, n_rotation, n_queries), sep='\n')


# T: O(n_queries)
# S: O(1) extra space
def queries(arr, n_rotation, n_queries):
    offset = n_rotation % len(arr)
    for _ in range(n_queries):
        query = morfo.read(int)
        yield arr[query - offset]


if __name__ == '__main__':
    main()

File: hackerrank/problem solving/algorithms/2 - implementation/26 - circular array rotation/with morfo library/circular_array_rotation.cs
// https://www.hackerrank.com/challenges/circular-array-rotation/problem?isFullScreen=true

using Morfo.IO;

var (n, nRotation, nQueries) = Reader.Read<int, int, int>();
int[] arr = Reader.Read<int>(n);
foreach (var q in Queries(arr, nRotation, nQueries))
    Console.WriteLine(q);

// T: O(nQueries)
// S: O(1) extra space
IEnumerable<int> Queries(int[] arr, int nRotation, int nQueries)
{
    int n = arr.Length;
    int offset = n - nRotation % n;
    for (int i = 0; i < nQueries; i++)
    {
        int query = Reader.Read<int>();
        int idx = (query + offset) % n;
        yield return arr[idx];
    }
}
